Source/CarthageKit/GitHub.swift

```diff
@@ -64,55 +64,144 @@ extension GitHubError: Decodable {
 	}
 }
 
-/// Describes a GitHub.com repository.
+/// Describes a GitHub.com or GitHub Enterprise repository.
 public struct GitHubRepository: Equatable {
+
+	/// Represents a GitHub server instance.
+	public enum Server: Equatable, Hashable, Printable {
+		/// The github.com server instance.
+		case GitHub
+
+		/// An Enterprise instance with its hostname.
+		case Enterprise(scheme: String, hostname: String)
+
+		public var scheme: String {
+			switch self {
+			case .GitHub:
+				return "https"
+
+			case let .Enterprise(scheme, _):
+				return scheme
+			}
+		}
+
+		public var hostname: String {
+			switch self {
+			case .GitHub:
+				return "github.com"
+
+			case let .Enterprise(_, hostname):
+				return hostname
+			}
+		}
+
+		public var APIEndpoint: String {
+			switch self {
+			case .GitHub:
+				return "\(scheme)://api.\(hostname)"
+
+			case .Enterprise:
+				return "\(description)/api/v3"
+			}
+		}
+
+		public var hashValue: Int {
+			return scheme.hashValue ^ hostname.hashValue
+		}
+
+		public var description: String {
+			return "\(scheme)://\(hostname)"
+		}
+	}
+
+	public let server: Server
 	public let owner: String
 	public let name: String
 
 	/// The URL that should be used for cloning this repository over HTTPS.
 	public var HTTPSURL: GitURL {
-		return GitURL("https://github.com/\(owner)/\(name).git")
+		return GitURL("\(server)/\(owner)/\(name).git")
 	}
 
 	/// The URL that should be used for cloning this repository over SSH.
 	public var SSHURL: GitURL {
-		return GitURL("ssh://git@github.com/\(owner)/\(name).git")
+		return GitURL("ssh://git@\(server.hostname)/\(owner)/\(name).git")
 	}
 
 	/// The URL for filing a new GitHub issue for this repository.
 	public var newIssueURL: NSURL {
-		return NSURL(string: "https://github.com/\(owner)/\(name)/issues/new")!
+		return NSURL(string: "\(server)/\(owner)/\(name)/issues/new")!
 	}
 
-	public init(owner: String, name: String) {
+	public init(server: Server = .GitHub, owner: String, name: String) {
+		self.server = server
 		self.owner = owner
 		self.name = name
 	}
 
-	/// Parses repository information out of a string of the form "owner/name".
-	public static func fromNWO(NWO: String) -> Result<GitHubRepository, CarthageError> {
-		let components = split(NWO, maxSplit: 1, allowEmptySlices: false) { $0 == "/" }
-		if components.count < 2 {
-			return .failure(CarthageError.ParseError(description: "invalid GitHub repository name \"\(NWO)\""))
+	/// Parses repository information out of a string of the form "owner/name"
+	/// for the github.com, or the form "http(s)://hostname/owner/name" for
+	/// Enterprise instances.
+	public static func fromIdentifier(identifier: String) -> Result<GitHubRepository, CarthageError> {
+		// GitHub.com
+		let components = split(identifier, maxSplit: 1, allowEmptySlices: false) { $0 == "/" }
+		if components.count == 2 && !components[0].hasPrefix("http") {
```

Comment:
You could use a regular expression to check for `^(\w+)/(\w+)$`—or code that checks for the equivalent.